### PR TITLE
[snapshot] Fixes failure on physical device due to snapshot()

### PR DESCRIPTION
**Pull request by umverteiler**
### Checklist
- [ ] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [ ] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Running any XCUITest that contains a snapshot() call on a **physical device** will fail. That's because the app property inside setupSnaphot() is set after (!) the setting of cacheDirectory. Unfortunately cacheDirectory can't be computed using pathPrefix() on a physical device due to the missing SIMULATOR_HOST_HOME property. So pathPrefix() will fail and by that always skip the setting of the app property as well (unnecessarily).

### Description
- Makes app and cacheDirectory properties in SnapshotHelper.swift optional (like they should be)
- Sets app before setting cacheDirectory
- On a physical device the snapshot() still not works, but at least it doesn't fails / crashes the test anymore

### CODE REVIEW

**Comment by LordNali** [comment]
I signed it!

**Comment by KrauseFx** [comment]
Hey @LordNali, thanks for your contribution on this. I didn't know you can run _snapshot_ on real devices, how does it work?

**Comment by LordNali** [comment]
You're right - you can't run snapshot on physical device - but we have some testcases that are using snapshot as a 'helper' to create screenshots for debugging / documentation during test execution. Theses testcases all fail on a physical device due to snapshot, even though the actual test would run perfectly on a physical device.

So in the end this fix doesn't help running snapshot on physical device, but prevent testcases that are using snapshot from crashing on physical devices ;-)

**Comment by KrauseFx** [comment]
Ah, alright, cool. Considering that, it probably doesn't make sense to merge this to the _fastlane_ code base, as it adds complexity, to cover a use-case we don't support. Is it cool to close this PR and you run your own fork for now?

**Comment by LordNali** [comment]
The implicit unwrapping of the properties app and cacheDirectory isn't good design. Furthermore setting property app only in case the cache directory has been set, doesn't seem to be good idea either.

I'm not sure what speaks against (merging) this PR, because in the end we're loosing nothing here, but will gain an error-free run in case someone runs tests that include a snapshot call on a physical device.

The purpose of the PR is not to enhance snapshot in terms of functionality, but to ensure that testcases that are using snapshot still run on a physical device.

File: snapshot/lib/assets/SnapshotHelper.swift [hunk 1]
```diff
@@ -80,6 +82,11 @@ open class Snapshot: NSObject {
     }
 
     class func setLanguage(_ app: XCUIApplication) {
+        guard let cacheDirectory = self.cacheDirectory else {
+            print("CacheDirectory is not set - probably runnig a physical device?")
```
Comment (KrauseFx) [inline snapshot/lib/assets/SnapshotHelper.swift]: `probably running on a physical device` <- added on

**Comment by KrauseFx** [comment]
Ahh, right, that makes sense. Sorry for asking stupid questions, I haven't actually learned Swift yet, so I appreciate you pointing out why we do this 👍

File: snapshot/lib/assets/SnapshotHelper.swift [hunk 2]
```diff
@@ -135,6 +152,12 @@ open class Snapshot: NSObject {
         #if os(OSX)
             XCUIApplication().typeKey(XCUIKeyboardKeySecondaryFn, modifierFlags: [])
         #else
+            
+            guard let app = self.app else {
+                print("XCUIApplication is not set")
```
Comment (KrauseFx) [inline snapshot/lib/assets/SnapshotHelper.swift]: This error message could be a little clearer so that the user knows what they have to do to fix the error 👍

**Comment by LordNali** [comment]
> ... I haven't actually learned Swift yet ...

And yet you're the author of one of the greatest tool(chain) for Swift Developers  :-)

I added the missing 'on' to the print message and extended the Snapshot error by a specific one for 'running on a physical device'. Finally I added the additional explanation in regard to fixing 'XCUIApplication is not set' error message.

File: snapshot/lib/assets/SnapshotHelper.swift [hunk 3]
```diff
@@ -83,7 +86,7 @@ open class Snapshot: NSObject {
 
     class func setLanguage(_ app: XCUIApplication) {
         guard let cacheDirectory = self.cacheDirectory else {
-            print("CacheDirectory is not set - probably runnig a physical device?")
+            print("CacheDirectory is not set - probably runnig on a physical device?")
```
Comment (janpio) [inline snapshot/lib/assets/SnapshotHelper.swift]: runnig => running

**Review by milch — COMMENTED** [review]
You also need to increment the version number in this file, see the comment at the beginning:

```
 // -----------------------------------------------------
 // IMPORTANT: When modifying this file, make sure to
 //            increment the version number at the very
 //            bottom of the file to notify users about
 //            the new SnapshotHelper.swift
 // -----------------------------------------------------
```

**Comment by LordNali** [comment]
Okay, updated it to 1.8.1 (assuming you're using some Semantic Versioning?).

**Comment by janpio** [comment]
Interesting, there seems to be a test for this that is now failing:

```
  1) Snapshot Snapshot::Runner Parses embedded SnapshotHelper.swift finds the current embedded version
     Failure/Error: expect(helper_version).to match(/^SnapshotHelperVersion \[\d.\d\]$/)

       expected "SnapshotHelperVersion [1.8.1]" to match /^SnapshotHelperVersion \[\d.\d\]$/
       Diff:
       @@ -1,2 +1,2 @@
       -/^SnapshotHelperVersion \[\d.\d\]$/
       +"SnapshotHelperVersion [1.8.1]"
     # ./snapshot/spec/runner_spec.rb:11:in `block (4 levels) in <top (required)>'
```

**Comment by LordNali** [comment]
Well, I'm not sure, but do we need to specify the new SnapshotHelper version anywhere else?

**Comment by janpio** [comment]
So I looked at this test a bit:
https://github.com/fastlane/fastlane/blob/c7b074dc4b564bea9f35814466734a704bcc20d4/snapshot/spec/runner_spec.rb#L6-L13
It seems to expect the version to be x.y, not x.y.z. So this indeed seems not to be semver, but Major.Minor. 

Solutions seems to be to change it to 1.9 then I guess?

File: snapshot/lib/assets/SnapshotHelper.swift [hunk 4]
```diff
@@ -135,6 +155,12 @@ open class Snapshot: NSObject {
         #if os(OSX)
             XCUIApplication().typeKey(XCUIKeyboardKeySecondaryFn, modifierFlags: [])
         #else
+            
+            guard let app = self.app else {
```
Comment (janpio) [inline snapshot/lib/assets/SnapshotHelper.swift]: Unrelated change?

**Review by janpio — APPROVED** [review]
Thanks, tests seem to like 1.9 :)

One last question, but I think this is good to merge no matter what you reply.

File: snapshot/lib/assets/SnapshotHelper.swift (hunk 4 above)
Comment (LordNali) [reply snapshot/lib/assets/SnapshotHelper.swift]: Not really - since property app has become optional (like cacheDirectory) we need to unwrap it before using it.

**Comment by janpio** [comment]
Question answered, so let's merge this 🥇 Thanks for the PR @LordNali.
